### ISSUE TITLE
Undocumented structures should show Swift declaration

### ISSUE THREAD

**Issue by juangamnik** [realm/jazzy#141]
Since #129 every undocumented structure has now a text "Undocumented" like here:
![jazzy_undoc](https://cloud.githubusercontent.com/assets/2939856/5631912/b8f4df0e-95ca-11e4-8444-c1121fa831e6.png)

But if I add some really meaningless documentation I get much more information (the Swift declaration):
![jazzy_decl](https://cloud.githubusercontent.com/assets/2939856/5631923/d191bf46-95ca-11e4-8876-6401c1d99e38.png)

So why not use the declaration instead of the meaningless notion "Undocumented" (I would just omit it).

**Comment by segiddins** [comment]
The issue here seems to be that SoureKit won't return any documentation XML when the declaration is undocumented. (See https://github.com/Realm/jazzy/blob/master/lib/jazzy/sourcekitten.rb#L97).

**Comment by juangamnik** [comment]
Uh. That's bad. But https://github.com/Realm/jazzy/blob/master/lib/jazzy/sourcekitten.rb#L96 (one line before) seems to access `key.doc.full_as_xml`. Perhaps there is another key, for directly accessing the declaration?

**Comment by juangamnik** [comment]
Anything on this? The "Undocumented"-text is unfortunately not very informative and the "Declaration" would be very helpful.

**Comment by jpsim** [comment]
I have not worked on this functionality lately, but you're of course welcome to do so. This part of jazzy's functionality is in SourceKitten.

**Comment by juangamnik** [comment]
@jpsim: I would, but unfortunately I'm neither familiar with ruby, nor with the project itself (from the source code point-of-view). Especially, if it is not an easy one (missing information from SourceKit as @segiddins said), which needs some digging in. My first approach would perhaps be to add some whitespace fake-documentation to each undocumented declaration (or even the words "not documented") in transient a preprocessing step ...

**Comment by jpsim** [comment]
SourceKitten is actually 100% written in Swift. You can find the repo here: https://github.com/jpsim/SourceKitten

**Comment by juangamnik** [comment]
Oh, cool. I'll look into it as soon as I can. Thanks.

**Comment by blochberger** [comment]
Interestingly if you have a documented class declaration and an undocumented member function, then the class declaration is shown as declaration for the member function, which is just erroneous:

```swift
///
public class Something {
	public func doSomething() {
		…
	}
}
```

The class documentation will render as:

> **Something**
>
>> ```swift
>> public class Something
>> ```

and the member function declaration of `doSomething()` will render as:

> Undocumented
>
> **Declaration**
>
>> ```swift
>> public class Something
>> ```

Adding an empty documentation explicitly will show the correct declaration, but will mess up documentation coverage:

```swift
///
public class Something {
	///
	public func doSomething() {
		…
	}
}
```

**Comment by istx25** [comment]
From my understanding, this bug is related to SourceKit which is out of our hands. I've tagged the issue with blocked and I'm going to close the issue. Feel free to reopen if you disagree or have more to say about it.

**Comment by jpsim** [comment]
Not exactly out of our hands, just more work!

**Comment by istx25** [comment]
Oh okay! Sweet, my apologies.

**Comment by juangamnik** [comment]
Phew, great. Was a bit anxious, that that's it (especially as I had no time yet to look into it and it doesn't look better in the near future)

**Comment by rcbello** [comment]
Hi, I am not sure if there is already a fix on the way for this issue but I am using this fix on my copy of jazzy.
In sourcekitten.rb, I modified `self.make.doc.info` to
````
 def self.make_doc_info(doc, declaration)
      return unless should_document?(doc)

      declaration.declaration = Highlighter.highlight(
        doc['key.parsed_declaration'] || doc['key.doc.declaration'],
        Config.instance.objc_mode ? 'objc' : 'swift',
      )
      if Config.instance.objc_mode && doc['key.swift_declaration']
        declaration.other_language_declaration = Highlighter.highlight(
          doc['key.swift_declaration'], 'swift'
        )
      end

      unless doc['key.doc.full_as_xml']
        return process_undocumented_token(doc, declaration)
      end
      ...
````
from
````
 def self.make_doc_info(doc, declaration)
      return unless should_document?(doc)

      unless doc['key.doc.full_as_xml']
        return process_undocumented_token(doc, declaration)
      end

      declaration.declaration = Highlighter.highlight(
        doc['key.parsed_declaration'] || doc['key.doc.declaration'],
        Config.instance.objc_mode ? 'objc' : 'swift',
      )
      if Config.instance.objc_mode && doc['key.swift_declaration']
        declaration.other_language_declaration = Highlighter.highlight(
          doc['key.swift_declaration'], 'swift'
        )
      end
     ...
````

I just moved the check and handling of undocumented structure to come after adding its declaration.
I have zero experience in Ruby so I'm putting it here to see if this is the right change or if this is the right place in the code to change.

**Comment by johnfairh** [comment]
Fixed in master via #897.

**Comment by juangamnik** [comment]
Thanks!
